x-pack/plugin/logsdb/src/test/java/org/elasticsearch/xpack/logsdb/SyntheticSourceLicenseServiceTests.java

```diff
@@ -8,54 +8,195 @@
+        when(licenseState.isAllowed(same(SyntheticSourceLicenseService.SYNTHETIC_SOURCE_FEATURE))).thenReturn(false);
+        licenseService.setLicenseState(licenseState);
+        licenseService.setLicenseService(mockLicenseService);
+        assertTrue("beyond cutoff date, so fallback to stored source", licenseService.fallbackToStoredSource(false, true));
+        Mockito.verify(licenseState, Mockito.never()).featureUsed(any());
+        Mockito.verify(licenseState, Mockito.times(1)).isAllowed(same(SyntheticSourceLicenseService.SYNTHETIC_SOURCE_FEATURE));
+    }
+
+    public void testGoldOrPlatinumLicenseCustomCutoffDate() throws Exception {
+        licenseService = new SyntheticSourceLicenseService(Settings.EMPTY, "2025-01-02T00:00");
+
+        long start = LocalDateTime.of(2025, 1, 1, 0, 0).toInstant(ZoneOffset.UTC).toEpochMilli();
+        License license = createGoldOrPlatinumLicense(start);
+        mockLicenseService = mock(LicenseService.class);
+        when(mockLicenseService.getLicense()).thenReturn(license);
+
+        MockLicenseState licenseState = MockLicenseState.createMock();
+        when(licenseState.getOperationMode()).thenReturn(license.operationMode());
+        when(licenseState.isAllowed(same(SyntheticSourceLicenseService.SYNTHETIC_SOURCE_FEATURE_LEGACY))).thenReturn(true);
+        licenseService.setLicenseState(licenseState);
+        licenseService.setLicenseService(mockLicenseService);
+        assertFalse("custom cutoff date, so fallback to stored source", licenseService.fallbackToStoredSource(false, true));
+        Mockito.verify(licenseState, Mockito.times(1)).featureUsed(any());
+        Mockito.verify(licenseState, Mockito.times(1)).isAllowed(same(SyntheticSourceLicenseService.SYNTHETIC_SOURCE_FEATURE_LEGACY));
+    }
+
+    static License createEnterpriseLicense() throws Exception {
+        long start = LocalDateTime.of(2024, 11, 12, 0, 0).toInstant(ZoneOffset.UTC).toEpochMilli();
+        return createEnterpriseLicense(start);
+    }
+
+    static License createEnterpriseLicense(long start) throws Exception {
+        String uid = UUID.randomUUID().toString();
+        long currentTime = System.currentTimeMillis();
+        final License.Builder builder = License.builder()
+            .uid(uid)
+            .version(License.VERSION_CURRENT)
+            .expiryDate(dateMath("now+2d", currentTime))
+            .startDate(start)
+            .issueDate(currentTime)
+            .type("enterprise")
+            .issuedTo("customer")
+            .issuer("elasticsearch")
+            .maxResourceUnits(10);
+        return TestUtils.generateSignedLicense(builder);
+    }
+
+    static License createGoldOrPlatinumLicense() throws Exception {
+        long start = LocalDateTime.of(2024, 11, 12, 0, 0).toInstant(ZoneOffset.UTC).toEpochMilli();
+        return createGoldOrPlatinumLicense(start);
+    }
+
+    static License createGoldOrPlatinumLicense(long start) throws Exception {
```

Comment:
Nit: remove this, pass the license string to the function above to deduplicate. Consider adding a boolean "isBeforeCutoff" to avoid passing dates.